src/abstractdataframe/join.jl

```diff
@@ -100,16 +100,33 @@ check_mapping_allowed(short::AbstractVector, refarray_long::AbstractVector,
     !isempty(short) && !isnothing(refpool_long) && !isnothing(invrefpool_long) &&
         eltype(refarray_long) <: Union{Signed, Unsigned}
 
-@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector) =
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{true}) =
+    [@inbounds mapping[r + 1] for r in refarray]
+
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{false}) =
     [@inbounds mapping[r] for r in refarray]
 
-map2refs(x::AbstractVector, invrefpool) = [get(invrefpool, v, nothing) for v in x]
+function map2refs(x::AbstractVector, invrefpool)
+    x_refpool = DataAPI.refpool(x)
+    x_refarray = DataAPI.refarray(x)
+    if isnothing(x_refpool) || !(x_refpool isa AbstractVector) || !(eltype(x_refarray) <: Integer)
+        return [get(invrefpool, v, nothing) for v in x]
+    else
+        # here we know that x_refpool is AbstractVector that allows integer indexing
+        # and its firstindex must be an integer
+        fi = firstindex(x_refpool)
+        # if fi is not 0 or 1 then we fallback to slow path for safety reasons
+        # all refpool we currently know have firstindex 0 or 1
+        # if there is some very strange firstindex we might run into oveflow issues
+        if 0 <= fi <= 1
+            mapping = [get(invrefpool, v, nothing) for v in x_refpool]
+            # use function barrier as mapping is type unstable
+            return map_refarray(mapping, x_refarray, Val(fi == 0))
```

Comment:
Why not simply pass `Val(fi)`?

Reply:
You could also pass `x_refpool` to `map_refarray` and do `+ firstindex(x_refpool)` there, since it's known statically.

Reply:
1. There is no guarantee that `firstindex` is known statically in general I think (it is possible that it might be impossible to be determined only by type of array)
2. I need to do `fi == 0` as I do not have a guarantee about type of `fi` - I only know it is `Integer`

If any of the above arguments is incorrect please comment, but I guess this is the cost we have to pay for having a very generic DataAPI.jl.

Reply:
Well at least you know that `0 <= fi <= 1`, so `Val(Int(fi))` is guaranteed to work. :-)

Reply:
Yes - I merged your proposal as I have checked that if we get `0` it is optimized out.